### PR TITLE
Adds next unstable cache for user subscription and welcome back message.

### PR DESCRIPTION
**Pull request by bLopata**
Adds unstable cache fetching for user subscription prior to chat inference. Also fixes 524 for showing `Welcome back` to a returning user. Refactors the existing supabase subscription queries/actions for determining chat access for a user and removes the `GET` operation from `admin`-scoped supabase client to follow permission of least privilege.

image

Sweet alert if user is not subscribed which navigates them to `settings/` to set up a subscription.
image

### CODE REVIEW

File: www/utils/supabase/queries.ts [hunk 1]
```diff
@@ -1,5 +1,42 @@
...
+export const getChatAccess = unstable_cache(
+  async (supabase: SupabaseClient, userId: string) => {
+    const { data: subscription } = await supabase
+      .from('subscriptions')
+      .select('status, trial_end')
+      .in('status', ['trialing', 'active'])
+      .maybeSingle();
+    
+    const isSubscribed = !!(subscription && subscription.status === 'active' && !subscription.trial_end);
+    const freeMessages = isSubscribed ? 0 : await getFreeMessageCount(userId);
```
Comment (VVoruganti) [inline www/utils/supabase/queries.ts:29]: Once you have the subscription already why do you need to do another call to getFreeMessageCount? This should already be in the metadata. You're repeating the same query multiple times here.

File: www/app/Chat.tsx [hunk 2]
```diff
@@ -257,28 +265,28 @@ export default function Chat({
...
+        confirmButtonColor: '#3085d6',
+        confirmButtonText: 'Subscribe',
+        showCancelButton: false,
+      }).then((result) => {
+        if (result.isConfirmed) {
+          router.push('/settings');
+        }
+      });
+      return;
+    }
```
Comment (VVoruganti) [inline www/app/Chat.tsx:283]: This is functionally doing the same thing as the previous method of checking for a valid subscription then checking for the number of free messages and not providing any additional protection. 

What you should be doing is moving the validation logic in to `app/api/chat/route.ts` and have it check for a valid subscription or remaining free messages.

This would still let someone reverse engineer the api and call the chat route from their developer tools.

File: www/app/page.tsx [hunk 3]
```diff
@@ -24,17 +23,23 @@ export default async function Home() {
...
   let freeMessages = 0;
+  let chatAccess;
 
   if (process.env.NEXT_PUBLIC_STRIPE_ENABLED === 'false') {
+    // In development, get the real message count but override subscription status
+    const realChatAccess = await getChatAccess(supabase, user.id);
+    chatAccess = {
+      isSubscribed: true,
+      freeMessages: realChatAccess.freeMessages,
+      canChat: realChatAccess.isSubscribed || realChatAccess.freeMessages > 0 // Always allow chat in development
```
Comment (VVoruganti) [inline www/app/page.tsx:34]: This comment is in-accurate and the value can still be false. This not always going to default to true

File: www/app/page.tsx [hunk 4]
```diff
@@ -24,17 +23,23 @@ export default async function Home() {
   // Get initial subscription state
   let isSubscribed = false;
   let freeMessages = 0;
+  let chatAccess;
```
Comment (VVoruganti) [inline www/app/page.tsx:26]: initialize this as an empty dictionary otherwise you might have weird behavior

File: www/app/page.tsx [hunk 5]
```diff
@@ -24,17 +23,23 @@ export default async function Home() {
...
-    isSubscribed = !!(sub && sub.status === 'active' && !sub.trial_end);
-
-    if (!isSubscribed) {
-      freeMessages = await getFreeMessageCount(user.id);
-    }
+    chatAccess = await getChatAccess(supabase, user.id);
+    isSubscribed = chatAccess.isSubscribed;
+    freeMessages = chatAccess.freeMessages;
+    // Set canChat based on subscription status or free message availability
+    chatAccess.canChat = chatAccess.isSubscribed || chatAccess.freeMessages > 0;
```
Comment (VVoruganti) [inline www/app/page.tsx:42]: Use the same format of assigning these at once in an object to maintain consistency

**Review by VVoruganti — CHANGES_REQUESTED** [review]
This approach isn't adding any extra protection and adds redundant queries. 

Please make changes to move validation logic into the correct location

File: www/utils/supabase/queries.ts [hunk 6]
```diff
@@ -1,5 +1,41 @@
...
+    const freeMessages = isSubscribed ? 0 : (subscription?.metadata as { freeMessages: number })?.freeMessages ?? 0;    
+    return {
+      isSubscribed,
+      freeMessages,
+      canChat: isSubscribed || freeMessages > 0
+    };
+  },
+  ['chat-access'],
+  { revalidate: 60 }
+);
```
Comment (VVoruganti) [inline www/utils/supabase/queries.ts:38]: Originally the getFreeMessageCount function would create a new trial subscription if one didn't exist, but by using this that call is never made and a trial subscription will never get created

File: www/utils/supabase/queries.ts [hunk 7]
```diff
@@ -1,5 +1,41 @@
...
+) => {
+  return cache(
+    next_unstable_cache(
+      callback as unknown as (...args: any[]) => Promise<Output>,
+      key,
+      options
+    )
+  );
+};
+
```
Comment (VVoruganti) [inline www/utils/supabase/queries.ts:19]: Should move this to its own utility file

File: www/utils/supabase/queries.ts (hunk 6 above)
Comment (VVoruganti) [reply www/utils/supabase/queries.ts:38]: Either call that if no subscription is made or refactor the logic to use that function instead

File: www/app/api/chat/route.ts [hunk 8]
```diff
@@ -204,10 +210,14 @@ export async function POST(req: NextRequest) {
...
     }
+    // Handle the stream response correctly
+    if (stream instanceof Response) {
+      return stream;
+    }
 
     console.log('Got the Stream');
 
-    return new NextResponse(stream.body, {
+    return new NextResponse(stream, {
```
Comment (VVoruganti) [inline www/app/api/chat/route.ts:220]: Why this change?

File: www/app/Chat.tsx [hunk 9]
```diff
@@ -336,9 +365,21 @@ export default function Chat({
...
+        // Convert ReadableStream to text first
+        const reader = honchoResponse.getReader();
+        let result = '';
+        while (true) {
+          const { done, value } = await reader.read();
+          if (done) break;
+          result += new TextDecoder().decode(value);
+        }
+        honchoContent = JSON.parse(result);
+      }
```
Comment (VVoruganti) [inline www/app/Chat.tsx:382]: We aren't streaming the Honcho response at this point so what does this all accomplish?

**Review by VVoruganti — CHANGES_REQUESTED** [review]

File: www/utils/supabase/queries.ts (hunk 7 above)
Comment (bLopata) [reply www/utils/supabase/queries.ts:19]: Gotcha.

File: www/app/api/chat/route.ts (hunk 8 above)
Comment (bLopata) [reply www/app/api/chat/route.ts:220]: Returning the 402 status code (payment required) in the route required changing the logic for response handling.

File: www/app/Chat.tsx (hunk 9 above)
Comment (bLopata) [reply www/app/Chat.tsx]: Was due to verbose http code check, fixed in latest commit.

File: www/app/api/chat/thought/route.ts
Comment (VVoruganti) [inline www/app/api/chat/thought/route.ts:1]: Fix merge conflicts

File: www/components/messagebox.tsx [hunk 10]
```diff
@@ -22,6 +22,7 @@ interface MessageBoxProps {
   userId?: string;
   conversationId?: string;
   message: Message;
+  chatError?: number;
```
Comment (VVoruganti) [inline www/components/messagebox.tsx:25]: This `chatError` is never used / no functionality ever populates the value

File: www/app/Chat.tsx [hunk 11]
```diff
@@ -305,13 +299,46 @@ export default function Chat({
...
+        setCanSend(false);
+        mutateMessages(
+          messages => [
+            ...(messages?.slice(0, -1) || []),
+            {
+              content: '',
+              isUser: false,
+              id: '',
+              metadata: {},
+              error: 402
```
Comment (VVoruganti) [inline www/app/Chat.tsx:318]: Should this be `chatError`?

File: www/app/Chat.tsx [hunk 12]
```diff
@@ -305,13 +299,46 @@ export default function Chat({
...
+          icon: 'warning',
+          confirmButtonColor: '#3085d6',
+          confirmButtonText: 'Subscribe',
+          showCancelButton: false,
+        }).then((result) => {
+          if (result.isConfirmed) {
+            router.push('/settings');
+          }
+        });
+        return;
```
Comment (VVoruganti) [inline www/app/Chat.tsx:335]: Don't need extra logic on front-end as the error should only occur on instances of users trying to bypass the UI

File: www/app/Chat.tsx [hunk 13]
```diff
@@ -55,6 +55,9 @@ async function fetchStream(
     });
 
     if (!response.ok) {
+      if (response.status === 402) {
+        return response;
```
Comment (VVoruganti) [inline www/app/Chat.tsx:59]: Shouldn't return anything in the case of an error just make a sweet alert directly

**Review by VVoruganti — CHANGES_REQUESTED** [review]

**Review by VVoruganti — APPROVED** [review]
